### PR TITLE
Add Go 1.20 support

### PR DESCRIPTION
**Pull request by austinvazquez**
* Add Go 1.20 to GitHub actions matrix
* Updates actions/setup-go from v1 to v3. (Resolves deprecation warnings in GitHub actions around Node 12. Reference [here](https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/)) 

Signed-off-by: Austin Vazquez <macedonv@amazon.com>

### CODE REVIEW

File: .github/workflows/build-pr.yml [hunk 1]
```diff
@@ -8,13 +8,18 @@ on:
 jobs:
   run:
     runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        go: ['1.18', '1.19', '1.20']
```
Comment (thaJeztah) [inline .github/workflows/build-pr.yml:14]: I'm always confused by the formats this action supports; is `1.18` equivalent to `1.18.0` or should it be `1.18.x`?

**Review by thaJeztah — APPROVED** [review]
LGTM (assuming the version format is correct)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
arf; looks like some changes are needed (see failures)

File: .github/workflows/build-pr.yml (hunk 1 above)
Comment (austinvazquez) [reply .github/workflows/build-pr.yml:14]: `1.18` should pull the latest; however, don't mind updating to `1.18.x` just to be more clear.

**Comment by austinvazquez** [comment]
> arf; looks like some changes are needed (see failures)

Added a little extra scope to resolve the errors. Took the opportunity to update the lint tooling from golint to golangci-lint and resolved new errors. Included here since we have the hood open but could split this into a seperate commit/PR if that helps.

**Review by tianon — APPROVED** [review]
Adding `go.mod` explicitly to `.gitignore` is a little odd when I don't think anything here generates it (right?), but I don't care about that enough to block this.

The other combined CI fixes/updates seem fine to include here IMO 👍

**Comment by austinvazquez** [comment]
> Adding `go.mod` explicitly to `.gitignore` is a little odd when I don't think anything here generates it (right?), but I don't care about that enough to block this.
> 
> The other combined CI fixes/updates seem fine to include here IMO 👍

Good call out I could have explained that better. CI generates them [here](https://github.com/opencontainers/runtime-spec/blob/main/.github/workflows/build-pr.yml#L25) and just wanted to be sure I didn't accidentally push those.

**Comment by AkihiroSuda** [comment]
Why not use go.mod ?

**Comment by austinvazquez** [comment]
> Why not use go.mod ?

I can make this happen. IIRC other specs had them.

**Review by AkihiroSuda — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,10 @@
+module github.com/opencontainers/runtime-spec
```
Comment (thaJeztah) [inline go.mod:1]: This has a potential wide impact than updating go versions in CI; do we still need to discuss;

https://github.com/opencontainers/runtime-spec/issues/1069

(or was there a concensus on that yet w.r.t. "SemVer from a Go perspective vs Spec perspective")?

Comment (AkihiroSuda) [reply go.mod:1]: image-spec and distribution-spec already have `go.mod`:
- https://github.com/opencontainers/image-spec/blob/main/go.mod
- https://github.com/opencontainers/distribution-spec/blob/main/specs-go/go.mod

And these struct definitions are unlikely to have breaking changes (until the spec itself has breaking changes)

Comment (AkihiroSuda) [reply go.mod:1]: @thaJeztah WDYT?

Comment (AkihiroSuda) [reply go.mod:1]: @austinvazquez @thaJeztah 
Do we need to split `go.mod` to another PR?

Comment (thaJeztah) [reply go.mod:1]: I would consider moving it separate yes, as it's not directly related to updating / updating CI to test against a newer go version

Comment (austinvazquez) [reply go.mod:1]: Okay, thanks for the feedback. Let me split and push.

**Review by tianon — APPROVED** [review]

**Review by kolyshkin — CHANGES_REQUESTED** [review]
Changes look good overall, but they need to be structured.

1.  Fix using deprecated ioutil.
2. Switch to using golangci-lint.
3. Modify ci to use matrix of go versions

One other thing, for golangci-lint it is better to use their action directly from .github/workflows rather than through make.

**Comment by AkihiroSuda** [comment]
@austinvazquez ↑ PTAL

**Comment by austinvazquez** [comment]
> Changes look good overall, but they need to be structured.
> 
> 1. Fix using deprecated ioutil.
> 2. Switch to using golangci-lint.
> 3. Modify ci to use matrix of go versions
> 
> One other thing, for golangci-lint it is better to use their action directly from .github/workflows rather than through make.

@kolyshkin, let me draft up what golangci-lint through github action would look like and you can lmk your thoughts.

**Comment by AkihiroSuda** [comment]
CI is failing

**Comment by austinvazquez** [comment]
> CI is failing

Oof I've fallen victim to the go.mod issue. Moved the hack up to resolve.

**Review by tianon — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**